packages/flutter_tools/lib/src/flutter_manifest.dart

```diff
@@ -33,13 +34,25 @@ class FlutterManifest {
     return _createFromYaml(loadYaml(manifest));
   }
 
-  static Future<FlutterManifest> _createFromYaml(Object yamlDocument) async {
+  static Future<FlutterManifest> _createFromYaml(dynamic yamlDocument) async {
     final FlutterManifest pubspec = new FlutterManifest._();
     if (yamlDocument != null && !await _validate(yamlDocument))
       return null;
 
-    pubspec._descriptor = yamlDocument ?? <String, dynamic>{};
-    pubspec._flutterDescriptor = pubspec._descriptor['flutter'] ?? <String, dynamic>{};
+    final Map<dynamic, dynamic> yamlMap = yamlDocument;
+    if (yamlMap != null) {
+      pubspec._descriptor = yamlMap.cast<String, dynamic>();
+    } else {
+      pubspec._descriptor = <String, dynamic>{};
+    }
+
+    final Map<dynamic, dynamic> flutterMap = pubspec._descriptor['flutter'];
+    if (flutterMap != null) {
+      pubspec._flutterDescriptor = flutterMap.cast<String, dynamic>();
+    } else {
+      pubspec._flutterDescriptor = <String, dynamic>{};
+    }
```

Comment:
Maybe extract this pattern to a facility like `castStringKeyedMap`. We could let that method return an empty map on `null` input (but then we might also want to rename it `toStringKeyedMap`).